api/oc_push.c

```diff
@@ -0,0 +1,2628 @@
+
+  OC_PROCESS_BEGIN();
+
+  while (1) {
+
+#if 0
+		int device_count = oc_core_get_num_devices();
+		/* create Push Notification Resource per each Device */
+		for (int i=0; i<device_count; i++) {
+			init_pushconf_resource(i);
+			init_pushreceiver_resource(i);
+		}
+#endif
+
+    OC_PROCESS_YIELD();
+
+    /* send UPDATE to target server */
+    if (ev == oc_events[PUSH_RSC_STATE_CHANGED]) {
+      ns_instance = (oc_ns_t *)data;
+      src_rsc = (oc_resource_t *)ns_instance->user_data;
+
+      if (!ns_instance || !src_rsc /*|| !ns_instance->user_data*/) {
+        OC_PUSH_ERR("something wrong! corresponding notification selector "
+                    "source resource is NULL, or updated resource is NULL!");
+        break;
+      }
+
+      /*
+       * client에서 POST 하는 루틴 참조할 것 (client_multithread_linux.c 참고)
+       */
+      /*
+       * 1. find `notification selector` which monitors `src_rsc` from
+       * `ns_col_list`
+       * 2. post UPDATE by using URI, endpoint (use oc_sting_to_endpoint())
+       */
+      if (oc_init_post(oc_string(ns_instance->targetpath),
+                       &ns_instance->pushtarget_ep, "if=oic.if.rw",
+                       &response_to_push_rsc, LOW_QOS, NULL)) {
+        /*
+         * add other properties than "rep" object of "oic.r.pushpayload"
+         * Resource here. payload_builder() only adds "rep" object.
+         *
+         * payload_builder() doesn't need to have "oc_rep_start_root_object()"
+         * and "oc_rep_end_root_object()" they should be added here...
+         */
+
+        oc_rep_begin_root_object();
+
+        /* anchor */
+        oc_uuid_to_str(oc_core_get_device_id(ns_instance->resource->device), di,
+                       OC_UUID_LEN);
+        oc_rep_set_text_string(root, anchor, di);
+
+        /* href (option) */
+        if (oc_string(ns_instance->phref) &&
+            strcmp(oc_string(ns_instance->phref), "")) {
+          oc_rep_set_text_string(root, href, oc_string(ns_instance->phref));
+        }
+
+        /* rt (array) */
+        oc_rep_open_array(root, rt);
+        for (size_t i = 0;
+             i < oc_string_array_get_allocated_size(src_rsc->types); i++) {
+          oc_rep_add_text_string(rt,
+                                 oc_string_array_get_item(src_rsc->types, i));
+        }
+        oc_rep_close_array(root, rt);
+
+        /* if (array) */
+        oc_core_encode_interfaces_mask(oc_rep_object(root),
+                                       src_rsc->interfaces);
+
+        src_rsc->payload_builder();
+
+        oc_rep_end_root_object();
+
+        if (oc_do_post())
+          OC_PUSH_DBG("Sent POST request");
+        else
+          OC_PUSH_ERR("Could not send POST");
+      } else {
+        OC_PUSH_ERR("Could not init POST");
+      }
+    }
+  }
+
+  OC_PROCESS_END();
+}
+
+/*
+ * @brief check if any of source array is part of target array
+ * @param target
+ * @param source
+ * @return
+ * 			0: any of source is not part of target
```

Comment:
this works but for `bool` it's natural to use true/false instead of 1/0

Reply:
fixed